builder/builder.go

```diff
@@ -381,8 +401,22 @@ func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt
 	return mts, nil
 }
 
+func (b *Builder) targetPhaseState(sts *states.SingleTarget) llb.State {
+	if b.builtMain {
+		return sts.RunPush.State
+	}
+	return sts.MainState
+}
+
+func (b *Builder) targetPhaseArtifacts(sts *states.SingleTarget) []states.SaveLocal {
+	if b.builtMain {
+		return sts.RunPush.SaveLocals
+	}
+	return sts.SaveLocals
+}
+
```

Comment:
Similar things to phases to enable use of the gwClient. However, its not quite the same as before where the state tracks it. Kept it simpler using the existing RunPush struct, and phases only happen here in builder where they are cared about now.

Reply:
Open to pushing this down and cleaning up some more; perhaps as a separate PR though? Its just touching a lot to do that

Reply:
Not sure what you have in mind by "cleaning up some more"?

This looks pretty good BTW.

Reply:
Cleaning up meaning pushing this down into the states package. Think of it as phases-lite; so the builder doesn't have to be so grabby with SingleTargets fields

Reply:
I think the way it is in this PR is simple enough. We could do that if things get more complex.